README.md
# 微信支付 API v3 Python SDK

[![PyPI version](https://badge.fury.io/py/wechatpayv3.svg)](https://badge.fury.io/py/wechatpayv3)
[![Download count](https://img.shields.io/pypi/dm/wechatpayv3)](https://img.shields.io/pypi/dm/wechatpayv3)

## 介绍

微信支付接口 V3 版 python 库。

欢迎微信支付开发者扫码进 QQ 群(群号：973102221)讨论，欢迎提交代码，欢迎star、follow、fork：

![image](qq.png)

## 适用对象

**wechatpayv3**同时支持微信支付[直连模式](https://pay.weixin.qq.com/wiki/doc/apiv3/index.shtml)及[服务商模式](https://pay.weixin.qq.com/wiki/doc/apiv3_partner/index.shtml)，接口说明详见官网。

## 特性

1. 平台证书自动更新，无需开发者关注平台证书有效性，无需手动下载更新；
2. 支持本地缓存平台证书，初始化时指定平台证书保存目录即可；
3. 敏感信息直接传入明文参数，SDK 内部自动加密，无需手动处理；
4. 回调通知自动验证回调消息，自动解密 resource 对象，并返回解密后的数据；
5. 已适配[直连模式](https://pay.weixin.qq.com/wiki/doc/apiv3/apis/index.shtml)和[服务商模式](https://pay.weixin.qq.com/wiki/doc/apiv3_partner/apis/index.shtml)中除“电商收付通”以外所有 v3 版接口。

## 源码

[github](https://github.com/minibear2021/wechatpayv3)

[gitee](https://gitee.com/minibear2021/wechatpayv3)

## 安装

```
$ pip install wechatpayv3
```

## 使用方法

### 准备

参考微信官方文档准备好密钥, 证书文件和配置([证书/密钥/签名介绍](https://pay.weixin.qq.com/wiki/doc/apiv3/wechatpay/wechatpay3_0.shtml))

- **商户 API 证书私钥：PRIVATE_KEY**。商户申请商户 API 证书时，会生成商户私钥，并保存在本地证书文件夹的文件 apiclient_key.pem 中。
  > :warning: 不要把私钥文件暴露在公共场合，如上传到 Github，写在客户端代码等。
- **商户 API 证书序列号：CERT_SERIAL_NO**。每个证书都有一个由 CA 颁发的唯一编号，即证书序列号。扩展阅读 [如何查看证书序列号](https://wechatpay-api.gitbook.io/wechatpay-api-v3/chang-jian-wen-ti/zheng-shu-xiang-guan#ru-he-cha-kan-zheng-shu-xu-lie-hao)。
- **微信支付 APIv3 密钥：APIV3_KEY**，是在回调通知和微信支付平台证书下载接口中，为加强数据安全，对关键信息 `AES-256-GCM` 加密时使用的对称加密密钥。

### 一个最小的后端

[examples.py](examples.py) 演示了一个带有[Native 支付下单](https://pay.weixin.qq.com/wiki/doc/apiv3/apis/chapter3_4_1.shtml)接口和[支付通知](https://pay.weixin.qq.com/wiki/doc/apiv3/apis/chapter3_4_5.shtml)接口的后端。
首先，修改 **examplys.py** 里以下几项配置参数：

```python
# 微信支付商户号，服务商模式下为服务商户号，即官方文档中的sp_mchid。
MCHID = '1230000109'

# 商户证书私钥，此文件不要放置在下面设置的CERT_DIR目录里。
with open('path_to_key/apiclient_key.pem') as f:
    PRIVATE_KEY = f.read()

# 商户证书序列号
CERT_SERIAL_NO = '444F4864EA9B34415...'

# API v3密钥， https://pay.weixin.qq.com/wiki/doc/apiv3/wechatpay/wechatpay3_2.shtml
APIV3_KEY = 'MIIEvwIBADANBgkqhkiG9w0BAQE...'

# APPID，应用ID，服务商模式下为服务商应用ID，即官方文档中的sp_appid，也可以在调用接口的时候覆盖。
APPID = 'wxd678efh567hg6787'

# 回调地址，也可以在调用接口的时候覆盖。
NOTIFY_URL = 'https://www.xxxx.com/notify'

# 微信支付平台证书缓存目录，初始调试的时候可以设为None，首次使用确保此目录为空目录。
CERT_DIR = './cert'

# 日志记录器，记录web请求和回调细节，便于调试排错。
logging.basicConfig(filename=os.path.join(os.getcwd(), 'demo.log'), level=logging.DEBUG, filemode='a', format='%(asctime)s - %(process)s - %(levelname)s: %(message)s')
LOGGER = logging.getLogger("demo")

# 接入模式：False=直连商户模式，True=服务商模式。
PARTNER_MODE = False

# 代理设置，None或者{"https": "http://10.10.1.10:1080"}，详细格式参见[https://requests.readthedocs.io/en/latest/user/advanced/#proxies](https://requests.readthedocs.io/en/latest/user/advanced/#proxies)
PROXY = None
```

接下来初始化 WechatPay 实例并配置一个合适的接口：

```python
wxpay = WeChatPay(
    wechatpay_type=WeChatPayType.NATIVE,
    mchid=MCHID,
    private_key=PRIVATE_KEY,
    cert_serial_no=CERT_SERIAL_NO,
    apiv3_key=APIV3_KEY,
    appid=APPID,
    notify_url=NOTIFY_URL,
    cert_dir=CERT_DIR,
    logger=LOGGER,
    partner_mode=PARTNER_MODE,
    proxy=PROXY)

app = Flask(__name__)

@app.route('/pay')
def pay():
    # 以native下单为例，下单成功后即可获取到'code_url'，将'code_url'转换为二维码，并用微信扫码即可进行支付测试。
    out_trade_no = ''.join(sample(ascii_letters + digits, 8))
    description = 'demo-description'
    amount = 1
    code, message = wxpay.pay(
        description=description,
        out_trade_no=out_trade_no,
        amount={'total': amount},
        pay_type=WeChatPayType.NATIVE
    )
    return jsonify({'code': code, 'message': message})
```

检查一下参数无误，现在就可以用 python 解释器来运行：

```shell
$ python examples.py
 * Serving Flask app "examples" (lazy loading)
 * Environment: production
   WARNING: This is a development server. Do not use it in a production deployment.
   Use a production WSGI server instead.
 * Debug mode: off
 * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)
```

现在访问 http://127.0.0.1:5000/pay ，如果一切正常，你会看到下面一串 json 字符串：

```python
{
  "code": 200,
  "message": "{\"code_url\":\"weixin://wxpay/bizpayurl?pr=abcdefghi\"}"
}
```

到这一步统一下单的后端就完成了，现在将 code_url 的值即"weixin://wxpay/bizpayurl?pr=abcdefghi"用[草料](https://cli.im/)转换为二维码即可用微信扫码进行支付测试。

**以上步骤如果不能正确执行，务必仔细检查各项初始化参数，必要的情况下，登录微信支付后台，将所有参数重置。**

Native 支付调试最简单便捷，调试通过没有问题证明初始化参数正确之后，如果需要采用其他（小程序、H5、JSAPI、APP）支付下单，可继续参考 examples.py。

## 接口清单

已适配的微信支付 V3 版 API 接口列表如下，部分接口调用示例可以参考[这里](interface.md)：

| 大类| 小类 | 接口 | 接口函数| 直连商户适用 | 服务商适用 |
| ---- | ---------------------------------------- | ------------------------ | --------------------------------------- | ------------ | ---------- |
| 公用| 公用 | 调起支付签名| sign | 是 | 是 |
| 公用| 公用 | 回调通知| callback | 是 | 是 |
| 公用| 公用 | 敏感信息参数解密 | decrypt | 是 | 是 |
| 公用| 公用 | 下载(交易、资金)账单| download_bill| 是 | 是 |
| 商户进件 | 特约商户进件、小微商户进件| 提交申请单 | applyment_submit | 否| 是|
| 商户进件 | 特约商户进件、小微商户进件| 查询申请单状态| applyment_query| 否| 是|
| 商户进件 | 特约商户进件、小微商户进件| 修改结算账号| applyment_settlement_modify| 否| 是|
| 商户进件 | 特约商户进件、小微商户进件| 查询结算账号| applyment_settlement_query| 否| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序支付| 统一下单 | pay | 是| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序支付| 查询订单 | query | 是| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序支付| 关闭订单 | close | 是| 是|
| 基础支付 | 合单支付 | 统一下单 | combine_pay| 是| 是|
| 基础支付 | 合单支付 | 查询订单 | combine_query| 是| 是|
| 基础支付 | 合单支付 | 关闭订单 | combine_close| 是| 是|
| 基础支付 | 退款 | 申请退款 | refund| 是| 是|
| 基础支付 | 退款 | 查询单笔退款| query_refund | 是| 是|
| 基础支付 | 退款 | 发起异常退款| abnormal_refund | 是| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序、合单支付 | 申请交易账单| trade_bill | 是| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序、合单支付 | 申请资金账单| fundflow_bill| 是| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序、合单支付 | 申请单个子商户资金账单 | submch_fundflow_bill| 否| 是|
| 基础支付 | JSAPI、APP、H5、Native、小程序、合单支付 | 下载账单 | download_bill| 是| 是|
| 经营能力 | 微信支付分（免确认模式） | 创单结单合并| payscore_direct_complete| 是| 否|
| 经营能力 | 微信支付分（免确认预授权模式）| 商户预授权 | payscore_permission | 是| 否|
| 经营能力 | 微信支付分（免确认预授权模式）| 查询用户授权记录| payscore_permission_query | 是| 否|
| 经营能力 | 微信支付分（免确认预授权模式）| 解除用户授权关系| payscore_permission_terminate| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 创建支付分订单| payscore_create| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 查询支付分订单| payscore_query | 是| 否|
| 经营能力 | 微信支付分（公共 API） | 取消支付分订单| payscore_cancel| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 修改订单金额| payscore_modify| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 完结支付分订单| payscore_complete| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 商户发起催收扣款| payscore_pay | 是| 否|
| 经营能力 | 微信支付分（公共 API） | 同步服务订单信息| payscore_sync| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 申请退款 | payscore_refund| 是| 否|
| 经营能力 | 微信支付分（公共 API） | 查询单笔退款| payscore_refund_query | 是| 否|
| 经营能力 | 微信支付分（公共 API） | 商户申请获取对账单| payscore_merchant_bill| 是| 否|
| 经营能力 | 支付即服务| 服务人员注册| guides_register| 是| 是|
| 经营能力 | 支付即服务| 服务人员分配| guides_assign| 是| 是|
| 经营能力 | 支付即服务| 服务人员查询| guides_query | 是| 是|
| 经营能力 | 支付即服务| 服务人员信息更新| guides_update| 是| 是|
| 经营能力 | 点金计划 | 点金计划管理| goldplan_plan_change| 否| 是|
| 经营能力 | 点金计划 | 商家小票管理| goldplan_custompage_change| 否| 是|
| 经营能力 | 点金计划 | 同业过滤标签管理| goldplan_advertising_filter| 否| 是|
| 经营能力 | 点金计划 | 开通广告展示| goldplan_advertising_open | 否| 是|
| 经营能力 | 点金计划 | 关闭广告展示| goldplan_advertising_close| 否| 是|
| 行业方案 | 电商收付通| 尚未适配 | 尚未适配| 否| 是|
| 行业方案 | 智慧商圈 | 商圈积分同步| points_notify| 是| 是|
| 行业方案 | 智慧商圈 | 商圈积分授权查询| user_authorization| 是| 是|
| 行业方案 | 智慧商圈 | 商圈会员待积分状态查询| business_point_status| 是| 是|
| 行业方案 | 智慧商圈 | 商圈会员停车状态同步| business_parking_sync| 是| 是|
| 行业方案 | 微信支付分停车服务| 查询车牌服务开通信息| parking_service_find| 是| 是|
| 行业方案 | 微信支付分停车服务| 创建停车入场| parking_enter| 是| 是|
| 行业方案 | 微信支付分停车服务| 扣费受理 | parking_order| 是| 是|
| 行业方案 | 微信支付分停车服务| 查询订单 | parking_query| 是| 是|
| 营销工具 | 代金券 | 创建代金券批次| marketing_favor_stock_create | 是| 是|
| 营销工具 | 代金券 | 激活代金券批次| marketing_favor_stock_start| 是| 是|
| 营销工具 | 代金券 | 发放代金券批次| marketing_favor_stock_send| 是| 是|
| 营销工具 | 代金券 | 暂停代金券批次| marketing_favor_stock_pause| 是| 是|
| 营销工具 | 代金券 | 重启代金券批次| marketing_favor_stock_restart| 是| 是|
| 营销工具 | 代金券 | 条件查询批次列表| marketing_favor_stock_list| 是| 是|
| 营销工具 | 代金券 | 查询批次详情| marketing_favor_stock_detail | 是| 是|
| 营销工具 | 代金券 | 查询代金券详情| marketing_favor_coupon_detail| 是| 是|
| 营销工具 | 代金券 | 查询代金券可用商户| marketing_favor_stock_merchant | 是| 是|
| 营销工具 | 代金券 | 查询代金券可用单品| marketing_favor_stock_item| 是| 是|
| 营销工具 | 代金券 | 根据商户号查用户的券| marketing_favor_user_coupon| 是| 是|
| 营销工具 | 代金券 | 下载批次核销明细| marketing_favor_use_flow| 是| 是|
| 营销工具 | 代金券 | 下载批次退款明细| marketing_favor_refund_flow| 是| 是|
| 营销工具 | 代金券 | 设置消息通知地址| marketing_favor_callback_update| 是| 是|
| 营销工具 | 商家券 | 创建商家券 | marketing_busifavor_stock_create | 是| 是|
| 营销工具 | 商家券 | 查询商家券详情| marketing_busifavor_stock_query| 是| 是|
| 营销工具 | 商家券 | 核销用户券 | marketing_busifavor_coupon_use | 是| 是|
| 营销工具 | 商家券 | 根据过滤条件查询用户券 | marketing_busifavor_user_coupon| 是| 是|
| 营销工具 | 商家券 | 查询用户单张券详情| marketing_busifavor_coupon_detail| 是| 是|
| 营销工具 | 商家券 | 上传预存 code | marketing_busifavor_couponcode_upload | 是| 是|
| 营销工具 | 商家券 | 设置商家券事件通知地址 | marketing_busifavor_callback_update| 是| 是|
| 营销工具 | 商家券 | 查询商家券事件通知地址 | marketing_busifavor_callback_query | 是| 是|
| 营销工具 | 商家券 | 关联订单信息| marketing_busifavor_coupon_associate| 是| 是|
| 营销工具 | 商家券 | 取消关联订单信息| marketing_busifavor_coupon_disassociate | 是| 是|
| 营销工具 | 商家券 | 修改批次预算| marketing_busifavor_stock_budget | 是| 是|
| 营销工具 | 商家券 | 修改商家券基本信息| marketing_busifavor_stock_modify | 是| 是|
| 营销工具 | 商家券 | 申请退券 | marketing_busifavor_coupon_return| 是| 是|
| 营销工具 | 商家券 | 使券失效 | marketing_busifavor_coupon_deactivate | 是| 是|
| 营销工具 | 商家券 | 营销补差付款| marketing_busifavor_subsidy_pay| 是| 是|
| 营销工具 | 商家券 | 查询营销补差付款单详情 | marketing_busifavor_subsidy_query| 是| 是|
| 营销工具 | 委托营销 | 建立合作关系| marketing_partnership_build| 是| 是|
| 营销工具 | 委托营销 | 查询合作关系列表| marketing_partnership_query| 是| 是|
| 营销工具 | 消费卡 | 发放消费卡 | marketing_card_send | 是| 否|
| 营销工具 | 支付有礼 | 创建全场满额送活动| marketing_paygift_activity_create| 是| 是|
| 营销工具 | 支付有礼 | 查询活动详情接口| marketing_paygift_activity_detail| 是| 是|
| 营销工具 | 支付有礼 | 查询活动发券商户号| marketing_paygift_merchants_list | 是| 是|
| 营销工具 | 支付有礼 | 查询活动指定商品列表| marketing_paygift_goods_list | 是| 是|
| 营销工具 | 支付有礼 | 终止活动 | marketing_paygift_activity_terminate| 是| 是|
| 营销工具 | 支付有礼 | 新增活动发券商户号| marketing_paygift_merchant_add | 是| 是|
| 营销工具 | 支付有礼 | 获取支付有礼活动列表| marketing_paygift_activity_list| 是| 是|
| 营销工具 | 支付有礼 | 删除活动发券商户号| marketing_paygift_merchant_delete| 是| 是|
| 营销工具 | 银行定向活促 | 出行券切卡组件预下单| industry_coupon_token| 否| 是|
| 营销工具 | 银行定向活促 | 导入定向用户协议号| bank_package_file| 否| 是|
| 营销工具 | 图片上传 | 图片上传(营销专用)| marketing_image_upload| 是| 是|
| 资金应用 | 商家转账到零钱 | 发起商家转账 | transfer_batch | 是| 否|
| 资金应用 | 商家转账到零钱 | 微信批次单号查询批次单| transfer_query_batchid | 是| 否|
| 资金应用 | 商家转账到零钱 | 微信明细单号查询明细单| transfer_query_detail_id | 是| 否|
| 资金应用 | 商家转账到零钱 | 商家批次单号查询批次单| transfer_query_out_batch_no | 是| 否|
| 资金应用 | 商家转账到零钱 | 商家明细单号查询明细单| transfer_query_out_detail_no | 是| 否|

CHANGELOG.md
# Changelog

## [1.2.41] - 2023-09-01

### Fixed

- 修正发起异常退款参数说明

## [1.2.40] - 2023-08-31

### Added

- 检查子商户开票功能状态
- 获取发票下载信息
- 下载发票文件
- 发起异常退款

## [1.2.39] - 2023-05-16

### Fixed

- 获取用户填写的抬头bug

## [1.2.38] - 2023-05-04

### Fixed

- 批次单号查询批次单bug

## [1.2.37] - 2023-04-02

### Added

- 查询结算账户修改申请状态

## [1.2.36] - 2023-02-24

### Added

- 导入定向用户协议号

## [1.2.35] - 2022-12-29

### Updated

- 发起批量转账接口增加transfer_scene_id参数

## [1.2.34] - 2022-09-23

### Fixed

- 下单接口pay_type未赋值异常

## [1.2.33] - 2022-09-20

### Added

- 所有下单接口新增请求参数字段： pay_type

## [1.2.32] - 2022-09-13

### Added

- 所有下单接口新增请求参数字段： support_fapiao

## [1.2.31] - 2022-09-01

### Added

- 商圈会员待积分状态查询
- 商圈会员停车状态同步

## [1.2.30] - 2022-07-20

### Added

- 出行券切卡组件预下单

### Fixed

- 修改商家券基本信息bug

## [1.2.29] - 2022-06-30

### Fixed

- 申请单个子商户资金账单bug

## [1.2.28] - 2022-06-16

### Updated

- 商户开户意愿确认-提交申请单参数调整
- 特约商户进件-提交申请单参数调整

## [1.2.27] - 2022-06-09

### Fixed

- 根据过滤条件查询用户券接口bug
- 修复电子发票国密SM3算法bug

## [1.2.26] - 2022-05-26

### Fixed

- 分账相关接口bug

## [1.2.25] - 2022-05-23

### Added

- 更新退款进度接口

## [1.2.24] - 2022-05-20

### Added

- 商家转账到零钱接口

## [1.2.23] - 2022-04-03

### Fixed

- 微信支付分请求退款接口bug

## [1.2.22] - 2022-04-07

### Fixed

- 请求分账回退bug

## [1.2.21] - 2022-04-06

### Fixed

- 创建代金券批次bug

## [1.2.20] - 2022-04-02

### Fixed

- 连锁品牌请求分账参数bug

## [1.2.19] - 2022-04-01

### Fixed

- 获取对私银行卡号开户银行加密bug

## [1.2.18] - 2022-04-01

### Added

- 银行组件接口

## [1.2.17] - 2022-03-23

### Fixed

- 修复请求分账接口加密参数bug

## [1.2.16] - 2022-03-14

### Fixed

- 修复开具电子发票时敏感信息未加密的bug

## [1.2.15] - 2022-03-14

### Added

- 支持电子发票相关接口

## [1.2.14] - 2022-03-07

### Fixed

- 修复商户进件提交申请单的bug

## [1.2.13] - 2022-02-17

### Added

- 初始化加载证书失败时抛出异常

## [1.2.12] - 2022-01-15

### Added

- 增加hmac_sha256签名方法，用于支持调起支付分订单时签名

### Fixed

- 创建支付分订单非必须参数值的 bug

## [1.2.11] - 2021-12-15

### Fixed

- notify_url 未赋值的 bug
- 连锁品牌请求分账的 bug

## [1.2.10] - 2021-12-05

### Fixed

- 请求分账的 bug

## [1.2.9] - 2021-12-05

### Fixed

- 请求分账的 bug

## [1.2.8] - 2021-12-05

### Fixed

- 查询代金券批次详情的一个 bug

## [1.2.7] - 2021-11-23

### Fixed

- 拼写错误

## [1.2.6] - 2021-11-20

### Added

- 新增 callback 接口，解密并保留所有回调参数，替代原 decrypt_callback 接口

## [1.2.5] - 2021-11-19

### Added

- 提交投诉回复新增可选参数
- 商户开户意愿确认接口
- 商户违规通知回调接口

## [1.2.4] - 2021-11-19

### Fixed

- 请求分账接口参数传递错误的 bug

## [1.2.3] - 2021-11-05

### Added

- 访问请求支持代理

## [1.2.2] - 2021-11-05

### Fixed

- 特约商户进件缺少 Wechatpay-Serial 的 bug

## [1.2.1] - 2021-11-04

### Fixed

- 拼写错误

## [1.2.0] - 2021-11-04

### Added

- 服务商模式接口(电商收付通除外)

### Fixed

- 直连模式部分接口 bug

## [1.1.0] - 2021-10-11

### Added

- 营销工具下代金券接口
- 营销工具下商家券接口

## [1.0.22] - 2021-10-09

### Fixed

- 查询支付分订单接口 bug

## [1.0.21] - 2021-09-29

### Added

- 商户申请获取对账单

## [1.0.20] - 2021-09-27

### Fixed

- 未引入支付分接口函数

## [1.0.19] - 2021-09-19

### Added

- 经营能力下微信支付分接口

### Changed

- 提前加载平台证书

## [1.0.18] - 2021-09-18

### Added

- 营销工具下支付有礼接口

## [1.0.17] - 2021-09-18

### Changed

- 日志信息内容

## [1.0.16] - 2021-09-18

### Added

- 初始化时支持日志记录

## [1.0.15] - 2021-09-17

### Added

- 营销工具下委托营销接口
- CHANGLOG.md

### Changed

- 示例 example

## [1.0.14] - 2021-09-14

### Changed

- 微调账单和图像下载接口返回值类型，message 参数从 string 变更为 bytes

## [1.0.13] - 2021-09-14

### Fixed

- 敏感信息解密 bug

## [1.0.12] - 2021-09-14

### Added

- 风险合规下消费者投诉 2.0 接口

## [1.0.11] - 2021-09-09

### Added

- 资金应用下分账接口

## [1.0.10] - 2021-09-05

### Changed

- 增强回调消息验证接口兼容性

## [1.0.9] - 2021-08-13

### Added

- 营销工具下图片上传(营销专用)接口

## [1.0.8] - 2021-08-13

### Added

- 行业方案下微信支付分停车接口

## [1.0.7] - 2021-08-12

### Added

- 其他能力下图片及视频上传接口

## [1.0.6] - 2021-08-09

### Changed

- 增强回调消息验证接口兼容性

## [1.0.5] - 2021-07-30

### Added

- 敏感信息参数自动加密

## [1.0.4] - 2021-07-14

### Added

- 经营能力下支付即服务接口

## [1.0.3] - 2021-06-22

### Fixed

- 下载证书无需验证签名

## [1.0.2] - 2021-06-16

### Fixed

- 平台证书更新 bug

## [1.0.1] - 2021-04-20

### Changed

- 代码重构

## [1.0] - 2021-04-20

### Added

- 智慧商圈积分授权查询

## [0.54] - 2021-04-20

### Added

- 支持本地缓存平台证书，减少网络请求

## [0.54] - 2021-04-20

### Changed

- 移除无效的 import

## [0.53] - 2021-04-19

### Changed

- 移除对 pyOpenSSL 包的依赖

## [0.53] - 2021-04-19

### Changed

- 移除对 pyOpenSSL 包的依赖

## [0.52] - 2021-04-19

### Added

- 行业方案下智慧商圈积分同步接口

### Fixed

- 签名验证 bug

## [0.51] - 2021-04-16

### Added

- 调起支付时需要的签名
- 回调消息接口自动解密 resource 字段

## [0.50] - 2021-04-16

### Added

- 微信平台证书自动下载及更新，无需提前准备

### Removed

- 证书下载接口

## [0.49] - 2021-04-16

### Changed

- 未提供平台证书时跳过接口验证

## [0.48] - 2021-04-16

### Fixed

- 验证签名 bug

## [0.47] - 2021-04-15

### Changed

- 接口返回类型从 bytes 更换为 string

## [0.46] - 2021-04-15

### Added

- 验证应答签名有效性

## [0.45] - 2021-04-15

### Fixed

- POST 接口签名错误

## [0.44] - 2021-04-15

### Added

- 基础支付下合单支付接口

## [0.43] - 2021-04-14

### Changed

- sdk 结构
- pypi 发布包格式

## [0.1] - 2021-04-14

### Added

- 基础支付(JSAPI 支付、APP 支付、H5 支付、Native 支付、小程序支付)下的：
  下单、查询订单、关闭订单、退款、退款查询、交易账单、资金账单及账单下载接口
- 新增证书下载接口


examples.py
# -*- coding: utf-8 -*-
import json
import logging
import os
from random import sample
from string import ascii_letters, digits

from flask import Flask, jsonify, request

from wechatpayv3 import SignType, WeChatPay, WeChatPayType

# 微信支付商户号（直连模式）或服务商商户号（服务商模式，即sp_mchid)
MCHID = '1234567890'

# 商户证书私钥
with open('path_to_key/apiclient_key.pem') as f:
    PRIVATE_KEY = f.read()

# 商户证书序列号
CERT_SERIAL_NO = '444F4864EA9B34415...'

# API v3密钥， https://pay.weixin.qq.com/wiki/doc/apiv3/wechatpay/wechatpay3_2.shtml
APIV3_KEY = 'MIIEvwIBADANBgkqhkiG9w0BAQE...'

# APPID，应用ID或服务商模式下的sp_appid
APPID = 'wxd678efh567hg6787'

# 回调地址，也可以在调用接口的时候覆盖
NOTIFY_URL = 'https://www.xxxx.com/notify'

# 微信支付平台证书缓存目录，减少证书下载调用次数，首次使用确保此目录为空目录.
# 初始调试时可不设置，调试通过后再设置，示例值:'./cert'
CERT_DIR = None

# 日志记录器，记录web请求和回调细节
logging.basicConfig(filename=os.path.join(os.getcwd(), 'demo.log'), level=logging.DEBUG, filemode='a', format='%(asctime)s - %(process)s - %(levelname)s: %(message)s')
LOGGER = logging.getLogger("demo")

# 接入模式:False=直连商户模式，True=服务商模式
PARTNER_MODE = False

# 代理设置，None或者{"https": "http://10.10.1.10:1080"}，详细格式参见https://docs.python-requests.org/zh_CN/latest/user/advanced.html
PROXY = None

# 初始化
wxpay = WeChatPay(
    wechatpay_type=WeChatPayType.NATIVE,
    mchid=MCHID,
    private_key=PRIVATE_KEY,
    cert_serial_no=CERT_SERIAL_NO,
    apiv3_key=APIV3_KEY,
    appid=APPID,
    notify_url=NOTIFY_URL,
    cert_dir=CERT_DIR,
    logger=LOGGER,
    partner_mode=PARTNER_MODE,
    proxy=PROXY)

app = Flask(__name__)


@app.route('/pay')
def pay():
    # 以native下单为例，下单成功后即可获取到'code_url'，将'code_url'转换为二维码，并用微信扫码即可进行支付测试。
    out_trade_no = ''.join(sample(ascii_letters + digits, 8))
    description = 'demo-description'
    amount = 1
    code, message = wxpay.pay(
        description=description,
        out_trade_no=out_trade_no,
        amount={'total': amount},
        pay_type=WeChatPayType.NATIVE
    )
    return jsonify({'code': code, 'message': message})


@app.route('/pay_jsapi')
def pay_jsapi():
    # 以jsapi下单为例，下单成功后，将prepay_id和其他必须的参数组合传递给JSSDK的wx.chooseWXPay接口唤起支付
    out_trade_no = ''.join(sample(ascii_letters + digits, 8))
    description = 'demo-description'
    amount = 1
    payer = {'openid': 'demo-openid'}
    code, message = wxpay.pay(
        description=description,
        out_trade_no=out_trade_no,
        amount={'total': amount},
        pay_type=WeChatPayType.JSAPI,
        payer=payer
    )
    result = json.loads(message)
    if code in range(200, 300):
        prepay_id = result.get('prepay_id')
        timestamp = 'demo-timestamp'
        noncestr = 'demo-nocestr'
        package = 'prepay_id=' + prepay_id
        paysign = wxpay.sign([APPID, timestamp, noncestr, package])
        signtype = 'RSA'
        return jsonify({'code': 0, 'result': {
            'appId': APPID,
            'timeStamp': timestamp,
            'nonceStr': noncestr,
            'package': 'prepay_id=%s' % prepay_id,
            'signType': signtype,
            'paySign': paysign
        }})
    else:
        return jsonify({'code': -1, 'result': {'reason': result.get('code')}})


@app.route('/pay_h5')
def pay_h5():
    # 以h5下单为例，下单成功后，将获取的的h5_url传递给前端跳转唤起支付。
    out_trade_no = ''.join(sample(ascii_letters + digits, 8))
    description = 'demo-description'
    amount = 1
    scene_info = {'payer_client_ip': '1.2.3.4', 'h5_info': {'type': 'Wap'}}
    code, message = wxpay.pay(
        description=description,
        out_trade_no=out_trade_no,
        amount={'total': amount},
        pay_type=WeChatPayType.H5,
        scene_info=scene_info
    )
    return jsonify({'code': code, 'message': message})


@app.route('/pay_miniprog')
def pay_miniprog():
    # 以小程序下单为例，下单成功后，将prepay_id和其他必须的参数组合传递给小程序的wx.requestPayment接口唤起支付
    out_trade_no = ''.join(sample(ascii_letters + digits, 8))
    description = 'demo-description'
    amount = 1
    payer = {'openid': 'demo-openid'}
    code, message = wxpay.pay(
        description=description,
        out_trade_no=out_trade_no,
        amount={'total': amount},
        pay_type=WeChatPayType.MINIPROG,
        payer=payer
    )
    result = json.loads(message)
    if code in range(200, 300):
        prepay_id = result.get('prepay_id')
        timestamp = 'demo-timestamp'
        noncestr = 'demo-nocestr'
        package = 'prepay_id=' + prepay_id
        paysign = wxpay.sign(data=[APPID, timestamp, noncestr, package], sign_type=SignType.RSA_SHA256)
        signtype = 'RSA'
        return jsonify({'code': 0, 'result': {
            'appId': APPID,
            'timeStamp': timestamp,
            'nonceStr': noncestr,
            'package': 'prepay_id=%s' % prepay_id,
            'signType': signtype,
            'paySign': paysign
        }})
    else:
        return jsonify({'code': -1, 'result': {'reason': result.get('code')}})


@app.route('/pay_app')
def pay_app():
    # 以app下单为例，下单成功后，将prepay_id和其他必须的参数组合传递给IOS或ANDROID SDK接口唤起支付
    out_trade_no = ''.join(sample(ascii_letters + digits, 8))
    description = 'demo-description'
    amount = 1
    code, message = wxpay.pay(
        description=description,
        out_trade_no=out_trade_no,
        amount={'total': amount},
        pay_type=WeChatPayType.APP
    )
    result = json.loads(message)
    if code in range(200, 300):
        prepay_id = result.get('prepay_id')
        timestamp = 'demo-timestamp'
        noncestr = 'demo-nocestr'
        package = 'Sign=WXPay'
        paysign = 'demo-sign'
        return jsonify({'code': 0, 'result': {
            'appid': APPID,
            'partnerid': MCHID,
            'prepayid': prepay_id,
            'package': package,
            'nonceStr': noncestr,
            'timestamp': timestamp,
            'sign': paysign
        }})
    else:
        return jsonify({'code': -1, 'result': {'reason': result.get('code')}})


@app.route('/notify', methods=['POST'])
def notify():
    result = wxpay.callback(request.headers, request.data)
    if result and result.get('event_type') == 'TRANSACTION.SUCCESS':
        resp = result.get('resource')
        appid = resp.get('appid')
        mchid = resp.get('mchid')
        out_trade_no = resp.get('out_trade_no')
        transaction_id = resp.get('transaction_id')
        trade_type = resp.get('trade_type')
        trade_state = resp.get('trade_state')
        trade_state_desc = resp.get('trade_state_desc')
        bank_type = resp.get('bank_type')
        attach = resp.get('attach')
        success_time = resp.get('success_time')
        payer = resp.get('payer')
        amount = resp.get('amount').get('total')
        # TODO: 根据返回参数进行必要的业务处理，处理完后返回200或204
        return jsonify({'code': 'SUCCESS', 'message': '成功'})
    else:
        return jsonify({'code': 'FAILED', 'message': '失败'}), 500


if __name__ == '__main__':
    app.run()
